### ISSUE TITLE
oci-distribution: not able to download standard images

### ISSUE THREAD

**Issue by gdesmott** [deislabs/krustlet#596]
To demontraste this problem I implemented a [simple pulling app](https://github.com/gdesmott/krustlet/blob/download-image/crates/oci-distribution/examples/download-image.rs).

Downloading a webasm image works fine:

```
$ RUST_LOG=oci_distribution=trace cargo run --example download-image -- webassembly.azurecr.io/hello-wasm
   Compiling oci-distribution v0.6.0 (/var/home/cassidy/dev/rust/krustlet/crates/oci-distribution)
    Finished dev [unoptimized + debuginfo] target(s) in 4.57s
     Running `/var/home/cassidy/dev/rust/krustlet/target/debug/examples/download-image webassembly.azurecr.io/hello-wasm`
[crates/oci-distribution/examples/download-image.rs:11] &reference = webassembly.azurecr.io/hello-wasm
[2021-05-11T13:24:18Z DEBUG oci_distribution::client] Pulling image: webassembly.azurecr.io/hello-wasm 
[2021-05-11T13:24:18Z DEBUG oci_distribution::client] Authorizing for image: webassembly.azurecr.io/hello-wasm 
[2021-05-11T13:24:18Z DEBUG oci_distribution::client] Making authentication call to https://webassembly.azurecr.io/oauth2/token 
[2021-05-11T13:24:19Z DEBUG oci_distribution::client] Received response from auth request: {"access_token":"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IjQ1NlA6WjNCRjpCQ0tPOkpUN0w6MlNLSTpVQUpZOkRCSU46VU5KWTpHWUZHOjdMQTI6WUpNSzpWUkVaIn0.eyJqdGkiOiI0MzA5OTk0OC01YWVmLTRiMzktODg5NC00YzZmMTBjYzEwMzUiLCJzdWIiOiJhY3JfYW5vbl9wdWxsIiwibmJmIjoxNjIwNzM4NTU4LCJleHAiOjE2MjA3Mzk3NTgsImlhdCI6MTYyMDczODU1OCwiaXNzIjoiQXp1cmUgQ29udGFpbmVyIFJlZ2lzdHJ5IiwiYXVkIjoid2ViYXNzZW1ibHkuYXp1cmVjci5pbyIsInZlcnNpb24iOiIxLjAiLCJyaWQiOiJjZWM4N2UxMzM4YjQ0MDY5YmI1ZGY3M2QxYmQxZTJjYiIsImFjY2VzcyI6W3sidHlwZSI6InJlcG9zaXRvcnkiLCJuYW1lIjoiKiIsImFjdGlvbnMiOlsicHVsbCJdfV0sInJvbGVzIjpbXSwiZ3JhbnRfdHlwZSI6ImFjY2Vzc190b2tlbiJ9.hcPn-XTPKI0-VELZvTJF6nwXXghZC1lqNMvU7BitQczj_sk4KxQvOvWbctfwQK7RDu0CtfS1nHQbTGJ7MHZbybAZGneBO-tAjzb0dw18lbaK3P7ZzSG_qnG2oYq1JWS10fy9agBlJBtNCeijdiy1ueXi6UIoNa1K_9aaMKs8MiNan-qG2H0qObs6xm5C9b_0yfkq_JnmcYTK5edDtpRva96c7RDHlS5HHuwt7cWxrRltAy83PAfPRLJ0ara7CPGZmbD7VdWTIxMJvKxWa1PGzFSGvD_sj3qC58m4-TtDEnPONYZHOCg1FNIIEqZWPp2JE0QiUTuOctr424_O7-oMsw"} 
[2021-05-11T13:24:19Z DEBUG oci_distribution::client] Succesfully authorized for image 'webassembly.azurecr.io/hello-wasm' 
[2021-05-11T13:24:19Z DEBUG oci_distribution::client] Pulling image manifest from https://webassembly.azurecr.io/v2/hello-wasm/manifests/latest 
[2021-05-11T13:24:19Z DEBUG oci_distribution::client] validating manifest: {"schemaVersion":2,"config":{"mediaType":"application/vnd.wasm.config.v1+json","digest":"sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a","size":2},"layers":[{"mediaType":"application/vnd.wasm.content.layer.v1+wasm","digest":"sha256:f9c91f4c280ab92aff9eb03b279c4774a80b84428741ab20855d32004b2b983f","size":1615998,"annotations":{"org.opencontainers.image.title":"module.wasm"}}]} 
[2021-05-11T13:24:19Z DEBUG oci_distribution::client] Parsing response as OciManifest: {"schemaVersion":2,"config":{"mediaType":"application/vnd.wasm.config.v1+json","digest":"sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a","size":2},"layers":[{"mediaType":"application/vnd.wasm.content.layer.v1+wasm","digest":"sha256:f9c91f4c280ab92aff9eb03b279c4774a80b84428741ab20855d32004b2b983f","size":1615998,"annotations":{"org.opencontainers.image.title":"module.wasm"}}]} 
[2021-05-11T13:24:19Z DEBUG oci_distribution::client] Pulling image layer 
Downloaded sha256:51d9b231d5129e3ffc267c9d455c49d789bf3167b611a07ab6e4b3304c96b0e7
```

But downloading the docker hello-world, also used in the [crate documentation](https://docs.rs/oci-distribution/0.6.0/oci_distribution/struct.Reference.html#examples), does not:

```
$ RUST_LOG=oci_distribution=trace cargo run --example download-image -- docker.io/library/hello-world:latest
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
     Running `/var/home/cassidy/dev/rust/krustlet/target/debug/examples/download-image 'docker.io/library/hello-world:latest'`
[crates/oci-distribution/examples/download-image.rs:11] &reference = docker.io/library/hello-world:latest
[2021-05-11T13:24:50Z DEBUG oci_distribution::client] Pulling image: docker.io/library/hello-world:latest 
[2021-05-11T13:24:50Z DEBUG oci_distribution::client] Authorizing for image: docker.io/library/hello-world:latest 
[2021-05-11T13:24:51Z DEBUG oci_distribution::client] Pulling image manifest from https://docker.io/v2/library/hello-world/manifests/latest 
Error: error decoding response body: expected value at line 2 column 1

Caused by:
    expected value at line 2 column 1
```

As the `oci-distribution` crate is meant to be usable for generic use cases I suppose there is something going wrong somewhere.

**Comment by gdesmott** [comment]
Interesting, using `registry-1.docker.io/library/hello-world:latest` instead of `docker.io/library/hello-world:latest` is working (except it's not because the crate does not support manifest list but that's another issue).

podman and docker accepts `docker.io/library/hello-world:latest` as image name, I'm not sure how, so it would be good if `oci-distribution` would as well.

**Comment by bacongobbler** [comment]
Right now we've been focused on pulling WebAssembly modules from an OCI registry, so we haven't focused much on the OCI image use case. If you find any issues, we'd welcome pull requests to fix those cases.

DockerHub does not support hosting "alternative" media types so we haven't been able to test against that endpoint yet.

**Comment by gdesmott** [comment]
> Do you know how Docker and Podman does the `docker.io` -> `registry-1.docker.io` conversion?

I don't speak Go but [looks like](https://github.com/containers/podman/blob/master/vendor/github.com/containers/image/v5/docker/docker_client.go#L249) it's hardcoded. Would it make sense to do the same in `oci-distribution` as well then?

**Comment by gdesmott** [comment]
I proposed a fix for this in #606
